osidb/serializer.py

```diff
@@ -635,6 +635,29 @@ def __init__(self, *args, **kwargs):
     def get_history(self, obj):
         """history events serializer getter"""
         history = pghistory.models.Events.objects.tracks(obj)
+        obj_id = (
+            self.context["request"].parser_context["kwargs"].get("id")
+            if "request" in self.context
+            else None
+        )
+
+        if obj_id:
+            for item in history:
+                if (
+                    item.pgh_context is not None
+                    and "user" in item.pgh_context
+                    and isinstance(item.pgh_context["user"], int)
+                ):
+                    user_id = item.pgh_context["user"]
+                    user = User.objects.get(id=user_id)
+                    item.pgh_context["user"] = (
+                        user.email or user.username or user.id or user
+                    )
+                    context_obj = pghistory.models.Context.objects.get(
+                        id=item.pgh_context_id
+                    )
+                    context_obj.metadata = item.pgh_context
+                    context_obj.save()
```

Comment:
I am not sure that change of the DB data in the serializer GET is a good approach. We are basically migrating the data on the fly.  Would it be possible to create a one time data migration instead?

Reply:
I am thinking about it this way. A migration is probably one time higher complexity task with a bit of risk of failing the deployment but then it is all over. A functionality like this is going to stay in the code base indefinitely and while it is probably less complicated now the complexity will stay with us potentially breaking any time in the future - something changes on the dependency update, someone makes a mistake in a related code etc.

If there is no strong reason why not to do this one time migration, I am really for it.

Reply:
I think we should definitely go with the migration approach here. Whatever mass change in data should be handled that way.

Reply:
Is it worth writing a backwards_func for this? it would store the user IDs mapped to the `pgh_context` entries

Reply:
I am not sure it is needed. I see we mostly do not write them. If it is simple then you can, otherwise there is probably no real reason to revert in the future. A **noop** should be enough.

Reply:
This on-the-fly-migrating code in the serializer should go away now when there is the migration.

Reply:
But we should instead somehow tweak the audit creation so it gets the user names/emails going forward.

Reply:
> But we should instead somehow tweak the audit creation so it gets the user names/emails going forward.

If understand your comment correctly, the middleware I wrote in #912 handles this

Reply:
You are right! I already forgot about this piece :facepalm: